File: QR-Encoder-Decoder/41vi4p/decode.py
import cv2
from pyzbar.pyzbar import decode
from PIL import Image
import sys
import io
import os

def decode_qr_to_image(input_qr, output_file):
    try:
        # Read the QR code image
        img = cv2.imread(input_qr)
        if img is None:
            raise Exception(f"Could not read image file: {input_qr}")
        
        # Decode QR code
        decoded_objects = decode(img)
        if not decoded_objects:
            raise Exception("No QR code found in image")
        
        for obj in decoded_objects:
            # Get the decoded data as bytes
            image_data = obj.data
            
            # Save the decoded data as image
            with open(output_file, 'wb') as f:
                f.write(image_data)
            print(f"Successfully decoded image to: {output_file}")
            return True
    except Exception as e:
        print(f"Error: {str(e)}")
        return False

def decode_qr_to_text(input_qr, output_file):
    try:
        # Read the QR code image
        img = cv2.imread(input_qr)
        if img is None:
            raise Exception(f"Could not read image file: {input_qr}")
        
        # Decode QR code
        decoded_objects = decode(img)
        if not decoded_objects:
            raise Exception("No QR code found in image")
        
        for obj in decoded_objects:
            # Get the decoded text
            text_data = obj.data.decode('utf-8')
            
            # Save the decoded text to file
            with open(output_file, 'w') as f:
                f.write(text_data)
            print(f"Successfully decoded text to: {output_file}")
            return True
    except Exception as e:
        print(f"Error: {str(e)}")
        return False

if __name__ == "__main__":
    # Decode first QR code to image
    if not decode_qr_to_image('img1.png', 'output1.png'):
        sys.exit(1)
        
    # Decode second QR code to text
    if not decode_qr_to_text('img2.png', 'output2.txt'):
        sys.exit(1)

File: QR-Encoder-Decoder/HGRgamer/decoder/decode.js
import { Jimp } from 'jimp';
import jsQR from 'jsqr';
import path from 'path';
import fs from 'fs';

function isbase64(data) {
    try {
        const decodedData = Buffer.from(data, 'base64');
        const encodedData = decodedData.toString('base64');
        return encodedData === data;
    } catch (e) {
        return false;
    }
}

function saveFile(decodedData, outputFile, isBinary) {
    const dir = path.dirname(outputFile);
    if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
    }

    if (isBinary) {
        const binaryArray = decodedData.split(' ').map(binary => parseInt(binary, 2));
        // Convert the array of integers into a Buffer (which holds binary data)
        const buffer = Buffer.from(binaryArray);
    
        fs.writeFileSync(outputFile, buffer);
        console.log(`Binary file saved: ${outputFile}`);
    } else {
        fs.writeFileSync(outputFile, decodedData, 'utf-8');
        console.log(`Text file saved: ${outputFile}`);
    }
}

async function decodeQRCode(imagePath) {
    try {
        const image = await Jimp.read(imagePath);

        const imageData = {
            data: new Uint8ClampedArray(image.bitmap.data),
            width: image.bitmap.width,
            height: image.bitmap.height,
        };

        const decodedQR = jsQR(imageData.data, imageData.width, imageData.height);
        if (!decodedQR) {
            throw new Error('QR code not found in the image.');
        }

        return decodedQR.data;
    } catch (error) {
        console.error('Error decoding QR code:', error);
        process.exit(1);
    }
}

async function processQRCode(imagePath, outputFile, isBinary) {
    try {
        const qrData = await decodeQRCode(imagePath);
        console.log('QR Code Data:', qrData);

        let decodedData;
        if (isbase64(qrData)) {
            decodedData = Buffer.from(qrData, 'base64').toString();
            console.log('Base64 Data Detected and Decoded.');
        } else {
            decodedData = qrData;
        }

        saveFile(decodedData, outputFile, isBinary);
    } catch (err) {
        console.error("Error:", err);
    }
}

var argv = process.argv.slice(2);
if(argv.length < 3) {
    console.log(`Usage: node decode.js <input file> <output file> <is binary(true/false)>`);
    process.exit(1);
}


const imagePath = argv[0];
const outputFile = argv[1];
const isBinary = (argv[2] == 'false') ? false : true;

// const imagePath = 'input1.png';
// const outputFile = 'output1.txt';
// const isBinary = false;

// const imagePath = 'input2.png';
// const outputFile = 'output2.png';
// const isBinary = true;

processQRCode(imagePath, outputFile, isBinary);